lib/matplotlib/colorbar.py

```diff
@@ -1250,7 +1275,12 @@ def update_normal(self, mappable):
             self.norm = mappable.norm
             self._reset_locator_formatter_scale()
 
-        self.draw_all()
+        try:
+            self.draw_all()
+        except AttributeError:
+            # update_normal sometimes is called when it shouldn't be..
```

Comment:
Why does `update_normal` get called when it shouldn't be? Presumably it has something to do with the new `cla()` addition you added. I wonder if you need to remove the `callbacks_cid` from `callbacksSM`?

Reply:
Oh more lovely callbacks ;-)

Reply:
The fundamental problem is that the colorbar axes is `cla`, but the `colorbar` object is not removed.  Then `Colorbar__init__` calls `mappable.autoscale_None()` which triggers a draw of the orphaned colorbar.   The orphaned colorbar tries to draw itself, but of course all its axes have been destroyed and we get an `AttributeError`.

Reply:
The problem is that for the callbacks, each mappable only maps to one colorbar, but in the following the mappable is used for two colorbars:

```python
import matplotlib.pyplot as plt 
import numpy as np 

fig, ax = plt.subplots()
pc = ax.imshow(np.arange(100).reshape(10, 10))
cb = fig.colorbar(pc, extend='both')
cb2 = fig.colorbar(pc)
# Clear and re-use the same colorbar axes
print('HERE1:', pc, pc.callbacksSM.callbacks)
cb.ax.cla()
print('HERE2:', pc, pc.callbacksSM.callbacks)
cb2.ax.cla()
print('HERE3:', pc, pc.callbacksSM.callbacks)
    
cb = fig.colorbar(pc, cax=cb.ax)
cb2 = fig.colorbar(pc, cax=cb2.ax)
```

which results in:

```
HERE1: AxesImage(80,52.8;317.44x369.6) {'changed': {0: <weakref at 0x1085d69e0; to 'Colorbar' at 0x1096758e0>, 1: <weakref at 0x109709120; to 'Colorbar' at 0x109748130>}}
id 1
HERE2: AxesImage(80,52.8;317.44x369.6) {'changed': {0: <weakref at 0x1085d69e0; to 'Colorbar' at 0x1096758e0>}}
id 1
HERE3: AxesImage(80,52.8;317.44x369.6) {'changed': {0: <weakref at 0x1085d69e0; to 'Colorbar' at 0x1096758e0>}}
```

For `HERE3` there should be no callback, but it never gets dicsonnected when we do `cb2.ax.cla()` because `pc` doesn't know about `cb2`.  

The culprit is here:

https://github.com/matplotlib/matplotlib/blob/d0137e9f1ef59f4fc1c1a81c61b04a573edfade2/lib/matplotlib/colorbar.py#L1181-L1183

Reply:
Of course this doesn't work at all on master?  So maybe this test is just too stringent.

Reply:
I guess this boils down to whether we want to properly support more than one colorbar per mappable or not.  The way the callbacks work, the answer is no; we don't forbid it, and it kind of works, but we can't remove the callbacks from the first axes in `cla()`.

Reply:
Here is an idea that would help this situation (and the pan/zoom on colorbars too which would help me). What about making the `Colorbar` object itself inherit from `CbarAxes`/become an Axes object? Then you would only call methods on the Colorbar object, not `cbar.ax`/`cax`...

Reply:
I don't know if that helps in particular.  

The issue here is that the scalar mappable is only storing a reference to a single colorbar, whereas it stores two callbacks.  So the line `self.colorbar.mappable.callbacksSM.disconnect(self.colorbar.mappable.colorbar_cid)` only knows about one colorbar id, and loses track of the other colorbar.  That leads to the stray draw.  

The newest commit promotes `colorbar_cid` to a dictionary if there is more than one and `.colorbar` to a list.  Thats a bit of a breaking change for what is stored on this attribute, but makes sure we don't lose track of what colorbars are on the scalar mappable.

Reply:
Oof, this opened a can of worms ;)

I don't think changing `colorbar_cid` to a dictionary is a great idea here, that seems like a lot more logic/work going on under the hood to take care of a niche case. As it so happens, it looks like there is some discussion about multiple callbacks in #20210 as well.

For this specific use-case, the reason you're getting multiple callbacks is because when you call `cbar.ax.cla()` that `ax.cla()` has no idea it should also clear a callback on the `cbar`. But, if we move that up a level to make it just `cbar.cla()` (i.e. Colorbar is now an Axes object), that clear call would know to get rid of the callback (and also clear the inner/outer axes as you already have implemented).

Reply:
I'm not at all fundamentally against that, but it seems a pretty hefty change. I think what you would do is lock out cb.ax and then just reproduce all the axes methods (which has been done already, but only partially). But it'd take some work to do it and not have a bunch of breaking API.

Reply:
Surprisingly, it doesn't seem that bad... See this commit:
https://github.com/greglucas/matplotlib/commit/569fbc1ae84e9a5b99ed938e9fd982a0c475f36d
Passes all `test_colorbar.py` tests locally. Just pointing `self.ax = self` after initialization would help and we could probably stick in a deprecation warning pass-through on that too. Maybe this is getting a little side-tracked for this thread and I should open up a separate issue for discussion?